README.md

```diff
@@ -48,6 +48,33 @@ the files (use the template above except for the `pre-commit` action):
         extra_args: flake8 --all-files
 ```
 
+### only check files that have changed
+
+`actions/checkout@v2` does a shallow clone by default which means pre-commit
+doesn't have enough Git history to determine which files have changed in order
+to only check them. So you need to expand the history yourself after the
+checkout action has run.
+
+This can be done effectively by getting a blobless clone with:
+```yaml
+    - uses: actions/checkout@v2
+    - name: Expand Git history
+      run: |
+        # Tell Git we want a blobless clone,
+        # which excludes file contents of historical commits making the cloning much faster.
+        git config remote.origin.partialclonefilter blob:none
```

Comment:
this doesn't work / persist -- the file blobs are still fetched when `pre-commit` internally runs `git diff --name-only`

Reply:
I'm not able to recreate that locally. I've tried with a couple of repositories and the size of the `.git` folder doesn't change after I have run `git diff --name-only`, even across spans of many years in large active repositories. The response is also basically instant from Git, which should indicate that it doesn't need to go and fetch extra data from the remote.

As a test I used the [Homebrew/brew](https://github.com/Homebrew/brew) repository.
* Doing a regular, full clone from scratch gave a `.git` folder of 50 MB.
* If I instead did a shallow clone (with `--depth=1`) it was only 3.3 MB.
* I then did the `git config [...]` and `git fetch --unshallow --no-tags` commands in the shallow repository, and that brought the size of `.git` up to 26 MB.
* Then I ran `git diff --name-only 1920eec...HEAD` (that's the SHA from the [2.0.0 tag](https://github.com/Homebrew/brew/releases/tag/2.0.0) from 2 years ago, and the size of the `.git` folder was still 26 MB.

I wonder if this might have been optimised in recent Git versions compared to when you may have tested it? I'm on version 2.30.1 FWIW.

Reply:
I just tried the same test on an Ubuntu 20.04 machine which has Git 2.25.1 available, and it didn't have to download the entire repository either.

Reply:
```console
$ git clone --depth 1 git@github.com:pre-commit/pre-commit
Cloning into 'pre-commit'...
remote: Enumerating objects: 266, done.
remote: Counting objects: 100% (266/266), done.
remote: Compressing objects: 100% (227/227), done.
remote: Total 266 (delta 13), reused 139 (delta 5), pack-reused 0
Receiving objects: 100% (266/266), 266.62 KiB | 2.12 MiB/s, done.
Resolving deltas: 100% (13/13), done.
$ cd pre-commit/
$ git config remote.origin.partialclonefilter blob:none
$ git fetch --unshallow --no-tags
remote: Enumerating objects: 6274, done.
remote: Counting objects: 100% (6274/6274), done.
remote: Compressing objects: 100% (2015/2015), done.
remote: Total 6238 (delta 4213), reused 6224 (delta 4200), pack-reused 0
Receiving objects: 100% (6238/6238), 947.13 KiB | 4.76 MiB/s, done.
Resolving deltas: 100% (4213/4213), completed with 11 local objects.
$ podman run --rm -ti --net=none -v $PWD:/src:rw --entrypoint sh docker.io/alpine/git
/git # cd /src/
/src # git diff HEAD^^^^^^^^^^^^^^^^ --name-only
ssh: Could not resolve hostname github.com: Try again
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
ssh: Could not resolve hostname github.com: Try again
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
fatal: unable to read d42bb1b1122b8c7e4216f91aa0e075f86bf006bd
```

Reply:
I'm not sure what problem you see with this. As far as I know, pre-commit doesn't work well without an internet connection either, luckily GitHub Actions does have access to the internet.

The suggestion to the README works for me, as I'm using it on a large private repository, where the options are:
* Use a shallow clone and run pre-commit across all the files
* Clone the entire repository and only run pre-commit across the changed files
* Set up this blobless clone and again just run across the changed files

Running pre-commit across all files take around 7-8 minutes, cloning the entire repository has to fetch 1.1 GB, so the third option works well for us.

Reply:
I'm demonstrating to you that it must pull objects to satisfy the `git diff --name-only` query